test/get-group-tangle.test.js

```diff
@@ -0,0 +1,237 @@
+  const server = Testbot({ name })
+
+  server.tribes2.create(null, (err, group) => {
+    if (err) throw err
+
+    const getGroupTangle = GetGroupTangle(server)
+
+    getGroupTangle(group.id, (err, groupTangle) => {
+      if (err) throw err
+
+      const { root, previous } = groupTangle
+      const rootKey = group.root
+
+      pull(
+        server.db.query(where(author(server.id)), descending(), toPullStream()),
+        pull.map((m) => m.key),
+        pull.take(1),
+        pull.collect((err, keys) => {
+          if (err) throw err
+
+          t.deepEqual(
+            { root, previous },
+            { root: rootKey, previous: [keys[0]] },
+            'group add-member of admin should be the tip'
+          )
+
+          //  publishing to the group:
+          const content = {
+            type: 'memo',
+            root: rootKey,
+            message: 'unneccessary',
+            recps: [group.id],
+          }
+
+          server.tribes2.publish(content, (err, msg) => {
+            if (err) throw err
+
+            getGroupTangle(group.id, (err, { root, previous }) => {
+              if (err) throw err
+              t.deepEqual(
+                { root, previous },
+                { root: rootKey, previous: [msg.key] },
+                'adding message to root'
+              )
+
+              server.tribes2.publish(content, (err, msg) => {
+                if (err) throw err
+
+                getGroupTangle(group.id, (err, { root, previous }) => {
+                  if (err) throw err
+                  t.deepEqual(
+                    { root, previous },
+                    { root: rootKey, previous: [msg.key] },
+                    'adding message to tip'
+                  )
+                  server.close(true, t.end)
+                })
+              })
+            })
+          })
+        })
+      )
+    })
+  })
+})
+
+const n = 100
+test(`get-group-tangle-${n}-publishes`, (t) => {
+  const publishArray = new Array(n).fill().map((item, i) => i)
+  const server = Testbot()
+  let count = 0
+  server.tribes2.create(null, (err, data) => {
+    if (err) throw err
+
+    const groupId = data.id
+    pull(
+      pull.values(publishArray),
+      paraMap(
+        (value, cb) =>
+          server.tribes2.publish({ type: 'memo', value, recps: [groupId] }, cb),
+        4
+      ),
+      paraMap((msg, cb) => server.db.getMsg(msg.key, cb), 10),
+      pull.drain(
+        (m) => {
+          count += m.value.content.tangles.group.previous.length
+        },
+        (err) => {
+          t.error(err, 'no error')
+
+          // t.equal(count, n, 'We expect there to be no branches in our groupTangle')
+          t.true(
+            count < n * 8,
+            'We expect bounded branching with fast publishing'
+          )
+
+          server.close(true, t.end)
+        }
+      )
+    )
+  })
+})
+
+test('get-group-tangle', (t) => {
+  const tests = [
+    {
+      plan: 4,
+      test: (t) => {
```

Comment:
What is this weird 4x test runner? Why do we need it?

Reply:
not sure, i think someone maybe tried to put all the tests into this one test at some point. but we could probably remove it

Reply:
Yeah let's remove anything that we don't know why we have it

Reply:
removed